### DISCUSSION TITLE
ListOf handling newline characters

### DISCUSSION THREAD

**Discussion by drush** [ohmjs/ohm#297] [Help] [answered]
Trying to figure out the syntax to allow lists to have support for multiple types of delimiters, ie comma and newline

```
one,two,three
```

```
one, two
three
```

```
one
two
three
```

This should fail (space is not a delimiter):
```
one two
```

Best code so far:
```
Table {
  Columns (a list of columns)
   = ListOf<ident, delim>
​
  delim  (an identifier)
    = "\n" | ","
    
  ident  (an identifier)
    = letter alnum*
}
```

But this doesn't recognize the newlines at all.  Changing the `ident` to `Ident` handles the newlines properly but doesn't reject the last case - it treats both words and the space as a single ident.  It should error.

**Comment by pdubroy** [comment, marked as answer]
Hi Darren,

If your language has significant newlines, it probably makes sense to redefine the `space` rule. That will change the behavior of the implicit space skipping. The [built-in definition](https://sourcegraph.com/github.com/harc/ohm/-/blob/ohm-js/src/Grammar.js#L342) includes any character from `\x00` up to and including `\x20` (space). E.g., you might want to define the `space` rule like this:

```
space := ~"\n" "\x00".."\x20"
```

Alternatively you could use `listOf` rather than `ListOf`, which would eliminate implicit space skipping, so you'd need to do it yourself. E.g., something like this may work:

```
Columns = listOf<ident, delim>
delim = spaceNoNL* ("\n" | ",")
spaceNoNL = ~"\n" "\x00".."\x20"
```

Does that help?

**Reply by drush** [reply]
This got us on the right track.  Thanks much!
